subprojects/logging/src/main/java/org/gradle/internal/logging/sink/GroupingProgressLogEventGenerator.java

```diff
@@ -0,0 +1,180 @@
+        if (operationId != null) {
+            operationsInProgress.get(operationId).bufferOutput(event);
+        } else {
+            onUngroupedOutput(event);
+        }
+    }
+
+    private void onComplete(ProgressCompleteEvent completeEvent) {
+        Object buildOpId = progressToBuildOpIdMap.get(completeEvent.getProgressOperationId());
+        if (operationsInProgress.containsKey(buildOpId)) {
+            operationsInProgress.remove(buildOpId).flushOutput();
+        }
+    }
+
+    private void onEnd(EndOutputEvent event) {
+        for (Map.Entry<Object, OperationGroup> entry : operationsInProgress.entrySet()) {
+            entry.getValue().flushOutput();
+        }
+        listener.onOutput(event);
+        buildOpIdHierarchy.clear();
+        operationsInProgress.clear();
+        progressToBuildOpIdMap.clear();
+    }
+
+    private void onUngroupedOutput(RenderableOutputEvent event) {
+        if (event.getBuildOperationId() != lastRenderedBuildOpId) {
+            listener.onOutput(new LogEvent(event.getTimestamp(), event.getCategory(), null, "", null));
+            lastRenderedBuildOpId = event.getBuildOperationId();
+        }
+        listener.onOutput(event);
+    }
+
+    // Return the id of the operation/group, checking up the build operation hierarchy
+    private Object getOperationId(@Nullable final Object buildOpId) {
+        Object current = buildOpId;
+        while (current != null) {
+            if (operationsInProgress.containsKey(current)) {
+                return current;
+            }
+            current = buildOpIdHierarchy.get(current);
+        }
+        return null;
+    }
+
+    private class OperationGroup {
+        private final Object buildOpIdentifier;
+        private final String category;
+        private final String loggingHeader;
+        private final long startTime;
+
+        private List<RenderableOutputEvent> bufferedLogs = new ArrayList<RenderableOutputEvent>();
+
+        private OperationGroup(String category, @Nullable String loggingHeader, long startTime, @Nullable Object buildOpIdentifier) {
+            this.category = category;
+            this.loggingHeader = loggingHeader;
+            this.startTime = startTime;
+            this.buildOpIdentifier = buildOpIdentifier;
+        }
+
+        StyledTextOutputEvent header(final String message) {
+            StyledTextOutputEvent.Span span = new StyledTextOutputEvent.Span(StyledTextOutput.Style.Header, "> " + message + EOL);
+            return new StyledTextOutputEvent(startTime, category, null, buildOpIdentifier, Collections.singletonList(span));
+        }
+
+        void bufferOutput(RenderableOutputEvent output) {
+            bufferedLogs.add(output);
+        }
+
+        void flushOutput() {
+            if (!bufferedLogs.isEmpty()) {
+                if (lastRenderedBuildOpId != buildOpIdentifier) {
+                    listener.onOutput(new LogEvent(startTime, category, null, "", null));
```

Comment:
The log event creation could be broken out into a separate method and reused above. Using a method would also _explain_ what this event represents.